src/script/api/script_log.cpp

```diff
@@ -18,75 +19,46 @@
 
 /* static */ void ScriptLog::Info(const char *message)
 {
-	ScriptLog::Log(LOG_INFO, message);
+	ScriptLog::Log(ScriptLogTypes::LOG_INFO, message);
 }
 
 /* static */ void ScriptLog::Warning(const char *message)
 {
-	ScriptLog::Log(LOG_WARNING, message);
+	ScriptLog::Log(ScriptLogTypes::LOG_WARNING, message);
 }
 
 /* static */ void ScriptLog::Error(const char *message)
 {
-	ScriptLog::Log(LOG_ERROR, message);
+	ScriptLog::Log(ScriptLogTypes::LOG_ERROR, message);
 }
 
-/* static */ void ScriptLog::Log(ScriptLog::ScriptLogType level, const char *message)
+/* static */ void ScriptLog::Log(ScriptLogTypes::ScriptLogType level, const char *message)
 {
-	if (ScriptObject::GetLogPointer() == nullptr) {
-		ScriptObject::GetLogPointer() = new LogData();
-		LogData *log = (LogData *)ScriptObject::GetLogPointer();
+	ScriptLogTypes::LogData &logdata = ScriptObject::GetLogPointer();
 
-		log->lines = CallocT<char *>(400);
-		log->type = CallocT<ScriptLog::ScriptLogType>(400);
-		log->count = 400;
-		log->pos = log->count - 1;
-		log->used = 0;
-	}
-	LogData *log = (LogData *)ScriptObject::GetLogPointer();
-
-	/* Go to the next log-line */
-	log->pos = (log->pos + 1) % log->count;
-
-	if (log->used != log->count) log->used++;
+	/* Limit the log to 400 lines. */
+	if (logdata.size() >= 400U) logdata.pop_front();
 
-	/* Free last message, and write new message */
-	free(log->lines[log->pos]);
-	log->lines[log->pos] = stredup(message);
-	log->type[log->pos] = level;
+	auto &line = logdata.emplace_back();
+	line.text = message;
```

Comment:
Maybe something like the following to not have to copy unneeded text? On the other hand, it's relatively unlikely for there to be newlines in the message so it might also be overkill/unneeded complication.
```
char *newline = strchr(message, '\n');
line.text = std::string(message, 0, newline == nullptr ? strlen(message) : newline - message);
```
Not that it matters much as with #10769 we'd be getting a `std::string&` as parameter, and then it's just `message.substr(0, message.find_first_of('\n'));`